dm/syncer/binlog_locations_test.go

```diff
@@ -0,0 +1,363 @@
+	s.currGSetStr = "3ccc475b-2343-11e7-be21-6c0b84d59f30:1-14"
+
+	var err error
+	s.prevGSet, err = gtid.ParserGTID(s.flavor, s.prevGSetStr)
+	c.Assert(err, IsNil)
+	s.lastGTID, err = gtid.ParserGTID(s.flavor, s.lastGTIDStr)
+	c.Assert(err, IsNil)
+	s.currGSet, err = gtid.ParserGTID(s.flavor, s.currGSetStr)
+	c.Assert(err, IsNil)
+
+	s.loc = binlog.Location{
+		Position: mysql.Position{
+			Name: s.binlogFile,
+			Pos:  s.binlogPos,
+		},
+	}
+	prevGSet := s.prevGSet.Origin()
+	c.Assert(s.loc.SetGTID(prevGSet), IsNil)
+}
+
+func (s *testLocationSuite) generateEvents(binlogEvents mockBinlogEvents, c *C) []*replication.BinlogEvent {
+	events := make([]*replication.BinlogEvent, 0, 1024)
+	for _, e := range binlogEvents {
+		switch e.typ {
+		case DBCreate:
+			evs, _, err := s.eventsGenerator.GenCreateDatabaseEvents(e.args[0].(string))
+			c.Assert(err, IsNil)
+			events = append(events, evs...)
+
+		case Write:
+			dmlData := []*event.DMLData{
+				{
+					TableID:    e.args[0].(uint64),
+					Schema:     e.args[1].(string),
+					Table:      e.args[2].(string),
+					ColumnType: e.args[3].([]byte),
+					Rows:       e.args[4].([][]interface{}),
+				},
+			}
+			eventType := replication.WRITE_ROWS_EVENTv2
+			evs, _, err := s.eventsGenerator.GenDMLEvents(eventType, dmlData)
+			c.Assert(err, IsNil)
+			events = append(events, evs...)
+
+		case DMLQuery:
+			dmlData := []*event.DMLData{
+				{
+					Schema: e.args[0].(string),
+					Query:  e.args[1].(string),
+				},
+			}
+			evs, _, err := s.eventsGenerator.GenDMLEvents(replication.UNKNOWN_EVENT, dmlData)
+			c.Assert(err, IsNil)
+			events = append(events, evs...)
+
+		case Headers:
+			filename := e.args[0].(string)
+			fakeRotate, err := utils.GenFakeRotateEvent(filename, uint64(s.binlogPos), s.serverID)
+			c.Assert(err, IsNil)
+			events = append(events, fakeRotate)
+
+			events1, _, err := event.GenCommonFileHeader(s.flavor, s.serverID, s.prevGSet)
+			c.Assert(err, IsNil)
+			events = append(events, events1...)
+
+		case Rotate:
+			nextFile := e.args[0].(string)
+			header := &replication.EventHeader{
+				Timestamp: uint32(time.Now().Unix()),
+				ServerID:  s.serverID,
+			}
+			e, err := event.GenRotateEvent(header, s.eventsGenerator.LatestPos, []byte(nextFile), 4)
+			c.Assert(err, IsNil)
+			events = append(events, e)
+		}
+	}
+	return events
+}
+
+// updateLastEventGSet increase the GTID set of last event.
+func (s *testLocationSuite) updateLastEventGSet(c *C, events []*replication.BinlogEvent) {
+	e := events[len(events)-1]
+	switch v := e.Event.(type) {
+	case *replication.XIDEvent:
+		v.GSet = s.currGSet.Origin()
+	case *replication.QueryEvent:
+		v.GSet = s.currGSet.Origin()
+	default:
+		c.Fatalf("last event is not expected, type %v", e.Header.EventType)
+	}
+}
+
+func (s *testLocationSuite) generateDMLEvents(c *C) []*replication.BinlogEvent {
+	var err error
+	s.eventsGenerator, err = event.NewGenerator(s.flavor, s.serverID, s.binlogPos, s.lastGTID, s.prevGSet, 0)
```

Comment:
seems `s. eventsGenerator` may be concurrently assigned, i am not sure if this will casuse unstable test